File: solutions/1544. Make The String Great/1544.py
# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
  def makeGood(self, s: str) -> str:
    ans = []
    for c in s:
      if ans and self.isBadPair(ans[-1], c):
        ans.pop()
      else:
        ans.append(c)

    return ''.join(ans)

  def isBadPair(self, a: str, b: str) -> bool:
    return a != b and a.lower() == b.lower()

File: solutions/1544. Make The String Great/1544.cpp
// Time complexity: O(n)
// Space complexity: O(n)
class Solution {
 public:
  string makeGood(string_view s) {
    string ans;
    for (const char c : s)
      if (!ans.empty() && isBadPair(ans.back(), c))
        ans.pop_back();
      else
        ans.push_back(c);

    return ans;
  }
 private:
  bool isBadPair(char a, char b) {
    return a != b && tolower(a) == tolower(b);
  }
};